MinnowBoard MRC based support

miczyg1 opened the pull request:
This PR adds silicon initialization not covered by native coreboot support but done in FSP. It also contains various fixes, TXE BIOS flow and CBFS_VERIFICATION integration with TXE Secure Boot on Bay Trail platform - Intel MinnowBoard Turbot.

miczyg1 committed:
soc/intel/baytrail: Put microcode and bootblock into BOOTBLOCK region

With CBFS_VERIFICATION one cannot fetch the early microcode update
from CBFS, as walkcbfs_asm is unsafe to be compiled with
CBFS_VERIFICATION. As a workaround, put a microcode copy in the IBB
under a fixed address. Bootblock and microcode will be verified by
TXE when TXE_SECURE_BOOT is enabled, so it will be safe to use.

Also put the microcode and bootblock in the BOOTBLOCK region which
will cover whole IBB to cleanly separate code and data to be verfied
by TXE from the rest of CBFS. it will make the updates easier and
keep CBFS_VERIFICATION working properly.

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
src/soc/intel/baytrail/northcluster.c: add NOOP set_resources to avoid an error in log

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
src/soc/intel/baytrail: Fix the FUNC_DIS bits description

BayTrail EDS is correct. MIPI is disabled in a different manner which
does not use FUNC_DIS register.

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
src/device/pciexp_device.c: Control ASPM with Kconfig option

Even if PCIEXP_ASPM is not selected, it doesn't mean all PCIe
devices will have the correct ASPM disabled state set. Ensure
the ASPM is disabled if it is not selected.

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
src/soc/intel/baytrail/Kconfig: Do not select PCIE_ASPM

Leave the choice to enable ASPM to the user via Kconfig.

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
soc/intel/baytrail/southcluster.c: Handle OTG and MIPI disabling properly

Disable OTG and MIPI as described in BayTrail BWG Vol. 2
(Document Number: 514148, Rev, 1.47).

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
src/soc/intel/baytrail/lpss.c: Configure INT_PIN as in FSP solution

Native code misses the interrupt configuration for LPSS devices.
Replicate the FSP default interrupt configuration.

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
soc/intel/baytrail/pcie.c: Add support for PCIe hotplug

By default FSP was configuring the ports as hotpluggable.
Allow hotplug configuration based on Kconfig. Use the
procedure described in Bay Trail BWG Vol.2 (Document
Number: 514148, Rev, 1.47) to initialize PCIe port.

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
src/soc/intel/baytrail/lpss.c: Add additional programming requirements per BWG

Implement additional LPSS programming requirements per Bay Trail BWG
Vol.2 (Document Number 514148, Rev, 1.47). These settings also reflect
the the FSP behavior.

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
soc/intel/baytrail/southcluster.c: Enable resources on LPC bridge

LPC bridge device must call enable_resources to recursively enable
resources for child devices. Otherwise child devices behind LPC bus
may have misconfigured resources.

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
soc/intel/baytrail: Add TXE BIOS flow implementation

Implement TXE BIOS flow as per Bay Trail TXE BWG
(Document Number: 514966, Revision 0.7).

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
device/pci_device.c: Clear bridge VGA16 decoding control bit after probing

After VGA16 bit was probed, it was never clear resulting in VGA16 bit
being enabled in all ports without any VGA device behind them. Clear
the VGA16 bit after probing and let coreboot set it back later for
devices where it is necessary.

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
src/mainboard/intel/minnowmax: Use SPDX license headers

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
payloads/external/SeaBIOS: Add option to override CBFS location

If the CBFS end is not at address 0xFFFFFFFF, then SeabIOS will not
be able to find CBFS. This option allows to set the CBFS end in
situations where COREBOOT region does not end at address 0xFFFFFFFF.

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
mb/intel/minnowmax: Add layout with separate bootblock and manifests

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
src/device/dram/ddr3.c: Add ECC type to DDR3 parsed data

The spd_add_smbios17 function already assumes MEMORY_BUS_WIDTH_64
non-ECC only, but did not fill in the ecc_type information, which
resulted in SMBIOS type 16 Memory Error Correction field to be out
of specitions (0 is not defined). Fill in the ecc_type field
as MEMORY_ARRAY_ECC_NONE.

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
southbridge/intel/common/firmware: Use the ME file path directly

Fixes a problem where changing ME binary in the config would not result
in the desired ME binary be included in the rebuilt coreboot binary.

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
soc/intel/baytrail: Change NC GPIO to input pull up 20K

All Intel SOCs use input pullup 20K as not connected GPIO.
Also GPIO output mode may be harmful.

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
mainboard/intel/minnowmax/devicetree.cb: Use default FSP settings

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
mainboard/intel/minnowmax/devicetree.cb: Enable both SATA ports

Minnowboard has two SATA ports: one on the standard SATA connector,
second on the expansion I/O header.

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
mainboard/intel/minnowmax/devicetree.cb: Describe OTG controller

Describe OTG controller in the devicetree. Also add comment about
known issue with writing to OTG IOSF port that will cause a hang
if OTG is disabled runtime using PMC FUNC_DIS. When OTG is disabled
by a soft strap in the flash descriptor, the issue does not occur
though.

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
3rdparty/dasharo-blobs: Bump For MinnowBoard blobs

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
configs/config.intel_minnowmax: Add configs for regular and SB build

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
arch/x86/smbios.c: Strip leading spaces from CPU brand string

Intel Bay Trail SoCs (confirmed on Atom E3845) have leading spaces
in the CPU brand string returned by CPUID. Strip these spaces so
that SMBIOS CPU string does not look weird in the EDK2 setup page
and the dmidecode output.

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
payloads/external/iPXE/Makefile: Use newer iPXE base revision

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 commented:
Needs this to land too: https://github.com/Dasharo/edk2/pull/156

miczyg1 commented on .github/workflows/build.yml:
```diff
@@ -192,6 +192,37 @@ jobs:
+    strategy:
+      matrix:
+        vendor: [ intel ]
+        model: [ minnowmax ]
```
Not sure what is wrong with this line. If anything it should complain on the next one

miczyg1 replied:
Added no_sb and sb prefixes, ugly, but works

miczyg1 committed:
build.sh: Add options to build Intel MinnowMax

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

krystian-hebel commented on src/soc/intel/baytrail/txei.c:
```diff
@@ -0,0 +1,1196 @@
+			 read32p(tempbar + TXE_MMIO_ALIVENESS_REQ) & (uint32_t)~1);
+
+	stopwatch_init_usecs_expire(&sw, HECI_READY_TIMEOUT);
+	while ((read32p(tempbar + TXE_MMIO_ALIVENESS_RESP) & request) != request) {
```
If `request == false`, this statement will never be true.

```suggestion
	while ((read32p(tempbar + TXE_MMIO_ALIVENESS_RESP) & 1) != request) {
```

Unless that is the expected behavior, in which case it should be described in the comment above the function.

krystian-hebel commented on src/soc/intel/baytrail/txei.c:
```diff
@@ -0,0 +1,1196 @@
+
+		udelay(1000);
+	} while(aliveness_req != aliveness_resp);
+
+	if ((read32p(tempbar + TXE_MMIO_ALIVENESS_REQ) & 1) == 1)
+		txe_aliveness_request(tempbar, false);
+
+	write32p(tempbar + TXE_MMIO_HOST_IPC_READINESS,
+		 read32p(tempbar + TXE_MMIO_HOST_IPC_READINESS) | IPC_READINESS_RDY_CLR);
+
+	stopwatch_init_usecs_expire(&sw, HECI_READY_TIMEOUT);
+	do {
+		if (stopwatch_expired(&sw)) {
+			printk(BIOS_ERR, "Timeout waiting for TXE IPC readiness\n");
+			return;
+		}
+
+		udelay(1000);
+	} while(!(read32p(tempbar + TXE_MMIO_SEC_IPC_READINESS) & IPC_READINESS_SEC_RDY));
+
+	write32p(tempbar + TXE_MMIO_HHISR,
+		 read32p(tempbar + TXE_MMIO_HHISR) & ~(HHISR_INT_BAR0_STS | HHISR_INT_BAR0_STS));
+
+	write32p(tempbar + TXE_MMIO_SEC_IPC_OUTPUT_STATUS,
+		 read32p(tempbar + TXE_MMIO_SEC_IPC_OUTPUT_STATUS) | IPC_OUTPUT_READY);
+
+
+	write32p(tempbar + TXE_MMIO_HOST_IPC_READINESS,
+		 read32p(tempbar + TXE_MMIO_HOST_IPC_READINESS) | IPC_READINESS_HOST_RDY);
```
This could use some comments.

miczyg1 replied:
I would add them if I knew what it really does and what for... I just barely managed to get it working...

krystian-hebel commented on src/soc/intel/baytrail/txei.c:
```diff
@@ -0,0 +1,1196 @@
+send_one_message(uint32_t hdr, const void *buff)
+{
+	size_t pend_len, pend_slots, remainder, i;
+	uint32_t tmp;
```
```suggestion
	uint32_t tmp = 0;
```

miczyg1 replied:
https://github.com/Dasharo/coreboot/pull/542/files#diff-9c3c7de7fe3776cfe7cd7ceaf5eaa33f7f23d275b737008f284189226f44ed2bR551

krystian-hebel commented on src/soc/intel/baytrail/txei.c:
```diff
@@ -0,0 +1,1196 @@
+	case TXE_FWSTS0_COM_DEBUG:
+	case TXE_FWSTS0_COM_SOFT_TEMP_DISABLE:
+	case TXE_FWSTS0_COM_SECOVER_JMPR:
+		printk(BIOS_INFO, "Current TXE operation mode does nto allow sending HECI messages\n");
```
```suggestion
		printk(BIOS_INFO, "Current TXE operation mode does not allow sending HECI messages\n");
```

krystian-hebel commented on src/soc/intel/baytrail/txei.c:
```diff
@@ -0,0 +1,1196 @@
+			continue;
+		}
+
+		cb_size = 16;
```
What is it? Why isn't this a constant? In fact, why `max_length` isn't a constant?

miczyg1 replied:
https://github.com/Dasharo/coreboot/pull/542/files#diff-9c3c7de7fe3776cfe7cd7ceaf5eaa33f7f23d275b737008f284189226f44ed2bR660
https://github.com/Dasharo/coreboot/pull/542/files#diff-9c3c7de7fe3776cfe7cd7ceaf5eaa33f7f23d275b737008f284189226f44ed2bR20-R21

krystian-hebel commented on src/soc/intel/baytrail/txei.c:
```diff
@@ -0,0 +1,1196 @@
+	printk(BIOS_DEBUG, "HECI: Global Reset Command\n");
+
+	if (!txe_cws_normal()) {
+		printk(BIOS_ERR, "HECI: TXE does not meet required prerequisites\n");
```
```suggestion
		printk(BIOS_ERR, "HECI: TXE does not meet prerequisites\n");
```

miczyg1 replied:
Based on common ME code: https://github.com/coreboot/coreboot/blob/master/src/soc/intel/common/block/cse/cse.c#L712

won't fix

krystian-hebel commented on src/soc/intel/baytrail/txei.c:
```diff
@@ -0,0 +1,1196 @@
+
+	status = heci_send_receive(&msg, sizeof(msg), &reply, &reply_size, HECI_MKHI_ADDR);
+
+	printk(BIOS_DEBUG, "HECI: Global Reset %s!\n", !status ? "success" : "failure");
```
In case of success HECI will reset the platform, right? In that case it doesn't make sense to print anything, it would be best to loop quietly until the reset happens.

miczyg1 replied:
Common ME code doesn't do it: https://github.com/coreboot/coreboot/blob/master/src/soc/intel/common/block/cse/cse.c#L727
and I won't either

pietrushnic commented:
@BeataZdunczyk @macpijan, could anyone fix the code according to Krystian's suggestion while @miczyg1 is on vacation? I also wonder if we shouldn't go upstream with those patches immediately.

BeataZdunczyk commented:
@pietrushnic, @miczyg1 is working today, so he will address the review and the issues raised from testing RC1. If anything is more challenging to address, we will, of course, assign the task to someone else. Upstream is out of scope for this release, but we can evaluate the effort needed.

miczyg1 commented on src/soc/intel/baytrail/Makefile.mk:
```diff
@@ -87,33 +93,26 @@ endif
-	dd if=/dev/zero of=$@ bs=5120 count=1 2> /dev/null
-	dd if=$< of=$@ conv=notrunc 2> /dev/null
-else
+key_manifest_pos = $(call int-subtract, $(CONFIG_ROM_SIZE) 0x21000)
```
I'm not sure I get it. What is wrong with the hardcoded size?

miczyg1 replied:
Ohh I get it, you mean to simply put the manifest using cbfstool without defining the lcoation.

filipleple replied:
@krystian-hebel so should we leave the hardcoded size?

krystian-hebel replied:
> MANIFESTS region may be placed anywhere...

Then it would be better to assert that this region is properly located, than blindly `dd` into offset that may be outside of MANIFESTS.

miczyg1 replied:
> Then it would be better to assert that this region is properly located, than blindly dd into offset that may be outside of MANIFESTS.

Yes, it could be checked somewhere. But blind DD is better than building an image which won't get the CPU out of reset...

miczyg1 replied:
Check implemented: a52993b1b61e111db94fa1b35574d9e7565a72a7
Also writing the key manifest with cbfstool using MANIFESTS region.

krystian-hebel replied:
It works, but is there a reason why there are two commas in `$(filter)`?

miczyg1 replied:
4a0e6c1f7e43e90fb061c1e1d924d4cc234fccce

No reason, probably some leftover from previous attempt to get this logic working...

miczyg1 commented on src/soc/intel/baytrail/txei.c:
```diff
@@ -0,0 +1,1196 @@
+		write32p(tempbar + TXE_MMIO_ALIVENESS_REQ,
+			 read32p(tempbar + TXE_MMIO_ALIVENESS_REQ) & (uint32_t)~1);
+
+	stopwatch_init_usecs_expire(&sw, HECI_READY_TIMEOUT);
+	while ((read32p(tempbar + TXE_MMIO_ALIVENESS_RESP) & request) != request) {
+		// Check if init timeout has expired.
+		if (stopwatch_expired(&sw)) {
+			printk(BIOS_ERR, "Timeout when sending TXE aliveness request: %d\n", request);
+			return -1;
+		}
+
+		udelay(100);
+	}
```
coreboot guys on the review love stopwatch.

krystian-hebel replied:
https://review.coreboot.org/c/coreboot/+/84862/1/src/soc/intel/common/block/cse/cse.c#368 are you sure about that? :slightly_smiling_face:

miczyg1 commented on src/soc/intel/baytrail/romstage/txe.c:
```diff
@@ -0,0 +1,274 @@
+	printk(BIOS_INFO, "Sending TXE DID message with status: %u\n", status >> 24);
+	pci_write_config32(txe, TXE_DID_MSG, status | DID_MSG_DID_DONE);
+
+	reg = pci_read_config32(txe, TXE_FWSTS0);
+	printk(BIOS_DEBUG, "TXE FWSTS0 before DID %08x\n", reg);
```
Not possible. There is no other code that will send DID. MRC binary does not do it.

miczyg1 commented on src/soc/intel/baytrail/romstage/romstage.c:
```diff
@@ -85,32 +86,34 @@ static int chipset_prev_sleep_state(const struct chipset_power_state *ps)
+
+	/* Get the UMA size */
+	uma_size = txe_get_uma_size();
+	mp.txe_size_mb = uma_size;
```
Setting TXE UMA size to 0 by MB will mean disabling TXE. I think I have explained it in another place in the code:
https://github.com/Dasharo/coreboot/pull/542/files#diff-c1b5561972dfba95e56690ad939598781658e9af5c4615242c673d2c611e250aR113

krystian-hebel replied:
Fixed: https://github.com/Dasharo/coreboot/commit/a18ce1e0318f391bacda6954dc8006637c9aea4b#diff-c3a5c5851544c283e1920ef54047754a2bf827a9c11f415cf821ae80eeb6802aL69

miczyg1 commented on src/soc/intel/baytrail/pmutil.c:
```diff
@@ -368,3 +370,34 @@ void poweroff(void)
+	if (!txe_global_reset())
+		return;
+
+	cf9_reset_prepare();
```
Ahh yes, the global_reste function does not need it indeed.

miczyg1 replied:
Removed. ALso added missing `halt()` after issuing the reset: https://github.com/Dasharo/coreboot/pull/542/files#diff-ca6a8eea6efbafef22d61dc25bb6e30e9d42849f4499ba6b0fed3614a5e74c70R399-R402

miczyg1 commented on src/soc/intel/baytrail/bootblock/microcode_asm.S:
```diff
@@ -0,0 +1,128 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
```
I decided not to modify the common file, because:

1. I would have to think how to compile the file with CONFIG_CBFS_VERIFICATION and avoid a situation when walkcbfs is not compiled and the code location is also not provided...
2. The code could use CPU_MICROCODE_CBFS_LOC but it is for CBFS, so won't work.
3. I would have to add my EARLY_CPU_MICROCODE_LOC

I didn't want to make it even more messy...

miczyg1 commented on src/soc/intel/baytrail/pcie.c:
```diff
@@ -81,10 +79,12 @@ static void byt_pcie_init(struct device *dev)
+#if !CONFIG(PCIEXP_HOTPLUG)
 		/* Disable hot plug, set power to 10W, set slot number. */
 		REG_PCI_RMW32(SLCAP, ~(HPC | HPS),
 			(1 << SLS_SHIFT) | (100 << SLV_SHIFT) |
 			(root_port_offset(dev) << SLN_SHIFT)),
```
I may have misunderstood BWG, because I though it gives me the order of operations to be done. SO if downstream device is not detected, the new should skip the slot power limit setting?

filipleple replied:
@krystian-hebel  IIUC as of [this commit](https://github.com/Dasharo/coreboot/pull/542/commits/bff6f1a2e42185d2a2f029412b6b56ed71849f04) hotplug is disabled if specified in the config file, whereas power and slot number get set regardless - not sure if I got it right it though

miczyg1 commented on src/soc/intel/baytrail/romstage/txe.c:
```diff
@@ -0,0 +1,274 @@
+		printk(BIOS_INFO, "TXE Current Operational Mode: SECOVER_JMPR\n");
+		txe_disable = true;
+		break;
+	case TXE_FWSTS0_COM_SECOVER_TXEI_MSG:
```
Yup, it should follow TXE disable path

miczyg1 replied:
https://github.com/Dasharo/coreboot/pull/542/files#diff-f23ab7181048cd01b365323ba7b06295c2e0703d0108de4c5f31f6c230de8624R130

miczyg1 commented on src/arch/x86/smbios.c:
```diff
@@ -39,6 +39,8 @@ static int smbios_processor_name(u8 *start)
 			}
 			tmp[12] = 0;
 			str = (const char *)tmp;
+			/* Skip leading spaces. */
```
Added https://github.com/Dasharo/coreboot/pull/542/commits/a62a04559046135da9d1a898d41369632d8448fd

miczyg1 commented on src/soc/intel/baytrail/Kconfig:
```diff
@@ -255,16 +259,39 @@ config TXE_SB_SBM_MANIFEST_SVN
-# Put the microcode right after the manifest, so it will be verified by IBB
-# and accessible without walkcbfs ASM routine if CBFS_VERIFICATION is enabled.
-config CPU_MICROCODE_CBFS_LOC
+# These addresses will be used by BSP to locate microcode in bootblock before
```
Fixed: https://github.com/Dasharo/coreboot/pull/542/files#diff-a68c92b0f440c50bce825bdbdc037e9897818b52b6545daa9255faf335b5c071R263

krystian-hebel replied:
Partially fixed, `addresses` now doesn't match `this`: https://github.com/Dasharo/coreboot/pull/542/files#diff-a68c92b0f440c50bce825bdbdc037e9897818b52b6545daa9255faf335b5c071R275

miczyg1 replied:
https://github.com/Dasharo/coreboot/pull/542/files#r1801179842

miczyg1 commented on src/soc/intel/baytrail/Kconfig:
```diff
@@ -255,16 +259,39 @@ config TXE_SB_SBM_MANIFEST_SVN
+# CBFS_VERIFICATION will be able to verify it already in ramstage. It is safe to
+# hardcode the address, because TXE will verify it together with IBB.
+
+config EARLY_CPU_MICROCODE_CBFS_LOC
```
Changed naming: https://github.com/Dasharo/coreboot/pull/542/files#diff-a68c92b0f440c50bce825bdbdc037e9897818b52b6545daa9255faf335b5c071R270

miczyg1 commented on src/soc/intel/baytrail/Makefile.mk:
```diff
@@ -87,33 +93,26 @@ endif
 
-manifests.bin-COREBOOT-position := $(call int-subtract, 0x100000000 0x21000)
+add_bootblock = \
+	$(CBFSTOOL) $(1) write -d -r BOOTBLOCK -f $(2); \
+	dd if=$(cpu_microcode_blob.bin-file) of=$(1) conv=notrunc bs=1 \
+		seek=$(microcode_pos) >/dev/null 2>&1
```
https://github.com/Dasharo/coreboot/pull/542/files#diff-a68c92b0f440c50bce825bdbdc037e9897818b52b6545daa9255faf335b5c071R171
https://github.com/Dasharo/coreboot/pull/542/files#diff-a68c92b0f440c50bce825bdbdc037e9897818b52b6545daa9255faf335b5c071R278-R284

miczyg1 commented on src/soc/intel/baytrail/romstage/txe.c:
```diff
@@ -0,0 +1,274 @@
+
+	/*
+	 * Add an extra 1*MiB memory for fTPM communication buffer.
+	 * MRC does allocates memory in MiB granualrity.
```
https://github.com/Dasharo/coreboot/pull/542/files#diff-f23ab7181048cd01b365323ba7b06295c2e0703d0108de4c5f31f6c230de8624R65

miczyg1 commented on src/soc/intel/baytrail/romstage/txe.c:
```diff
@@ -0,0 +1,274 @@
+	}
+
+	if ((reg & DID_MSG_ACK_MASK) != DID_MSG_GOT_ACK) {
+		printk(BIOS_ERR, "TImeout wwaiting for TXE DRAM Init Done ACK!\n");
```
https://github.com/Dasharo/coreboot/pull/542/files#diff-f23ab7181048cd01b365323ba7b06295c2e0703d0108de4c5f31f6c230de8624R184

miczyg1 commented on src/soc/intel/baytrail/romstage/txe.c:
```diff
@@ -0,0 +1,274 @@
+
+	reg &= DID_ACK_DATA_MASK;
+	printk(BIOS_DEBUG, "TXE BIOS action: %u\n", reg >> DID_ACK_DATA_SHIFT);
+	/* Ensure resources are programmed before requestign global reset */
```
https://github.com/Dasharo/coreboot/pull/542/files#diff-f23ab7181048cd01b365323ba7b06295c2e0703d0108de4c5f31f6c230de8624R190

miczyg1 commented on src/soc/intel/baytrail/romstage/txe.c:
```diff
@@ -0,0 +1,274 @@
+		pci_read_config32(txe, TXE_SATT1_BIOS_CTRL));
+
+	txe_write_did_message(txe, DID_STATUS_SUCCESS);
+}
```
Fixed: https://github.com/Dasharo/coreboot/pull/542/files#diff-f23ab7181048cd01b365323ba7b06295c2e0703d0108de4c5f31f6c230de8624R275

miczyg1 commented on src/soc/intel/baytrail/txe.c:
```diff
@@ -0,0 +1,71 @@
+	heci_bar1 = find_resource(dev, PCI_BASE_ADDRESS_1);
+
+	if (!heci_bar1 || heci_bar1->base == 0) {
+		printk(BIOS_DEBUG, "HECI BAR1 not assigned, not sending aliveness request");
```
https://github.com/Dasharo/coreboot/pull/542/files#diff-1bf7e5734b10baca8f1ca18931620c2bbb0b94672e56402d9b38d5b8a652eab2R52

miczyg1 commented on src/soc/intel/baytrail/txei.c:
```diff
@@ -0,0 +1,1196 @@
+bool is_txe_devfn_visible(void)
+{
+	if (!is_devfn_enabled(PCI_DEVFN(TXE_DEV, TXE_FUNC))) {
+		printk(BIOS_WARNING, "TXE device 1a.0 is disabled\n");
```
https://github.com/Dasharo/coreboot/pull/542/files#diff-9c3c7de7fe3776cfe7cd7ceaf5eaa33f7f23d275b737008f284189226f44ed2bR43

miczyg1 commented on src/soc/intel/baytrail/txei.c:
```diff
@@ -0,0 +1,1196 @@
+	}
+
+	if (pci_read_config16(PCI_DEV(0, TXE_DEV, TXE_FUNC), PCI_VENDOR_ID) == 0xFFFF) {
+		printk(BIOS_WARNING, "TXE device 1a.0 is hidden\n");
```
https://github.com/Dasharo/coreboot/pull/542/files#diff-9c3c7de7fe3776cfe7cd7ceaf5eaa33f7f23d275b737008f284189226f44ed2bR48

miczyg1 commented on src/soc/intel/baytrail/southcluster.c:
```diff
@@ -415,6 +416,42 @@ static void sc_init(struct device *dev)
+	u32 val;
+
+	/*
+	 * ISP/IUNIT PCI D3hot request via PCI PM_CAP is ignoreed by HW.
```
https://github.com/Dasharo/coreboot/pull/542/files#diff-f4c302b68b226b0d9c61f6147dadfe9e541b0f755ff685d7abe1295245814cf9R424

miczyg1 commented:
I have addressed what I could. Not sure how to address the PCIe comments properly yet.

filipleple committed:
src/device/pciexp_device.c: switch ASPM disabling order

Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple commented on src/device/pciexp_device.c:
```diff
@@ -586,9 +586,22 @@ static void pciexp_enable_aspm(struct device *root, unsigned int root_cap,
+		lnkctl &= ~3;
+		pci_write_config16(root, root_cap + PCI_EXP_LNKCTL, lnkctl);
+
+		/* Disable ASPM in endpoint device next */
```
[Reversed](https://github.com/Dasharo/coreboot/pull/542/commits/208723b73b644dce7cf5055ca4c095d16f65c970) the order.

miczyg1 replied:
4ae754bc80be096ead91ffd846fa20a64557dc80 
Made disable_pcie_aspm to disable ASPM too.

filipleple committed:
src/soc/intel/baytrail/pcie.c: set power and slot number regardless of hotplug

Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple committed:
src/soc/intel/baytrail/pcie.c: update hotplug function

Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple commented on src/soc/intel/baytrail/pcie.c:
```diff
@@ -170,26 +173,45 @@ static u8 all_ports_no_dev_present(struct device *dev)
+	struct reg_script hotplug_port[] = {
+		REG_PCI_RMW32(PCIEALC, ~(1 << 26), 0),
+		/* Configure hot plug, set power to 10W, set slot number. */
+		REG_PCI_RMW32(SLCAP, ~(HPC | HPS),
```
I did apply the previous suggestion, so I used the one-liner in [this commit](https://github.com/Dasharo/coreboot/pull/542/commits/55ba4b3df4154799d1c6c1b25c637aa237b67e00)

filipleple committed:
src/mainboard/intel/minnowmax/Kconfig: enable S3 suspend

Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

miczyg1 committed:
src/mainboard/intel/minnowmax/devicetree.cb: Enable HD Audio

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
src/soc/intel/baytrail: Publish device NVS to avoid ACPI errors

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
src/soc/intel/baytrail/cpu.c: Sync MTRRs in later boot phase

Linux complained on inconsistent MTRRs. Apparently they were not set for
APs correctly.

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
soc/intel/baytrail: Add option to disable HPET

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
soc/intel/baytrail: Add proper UART support

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
soc/intel/baytrail/northcluster.c: Do not report the same region twice

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
src/soc/intel/baytrail/Kconfig: Disable S1

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
configs/config.intel_minnowmax_no_sb: Disable SMM BWP

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
acpi: Add support for CSRT table

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
soc/intel/baytrail/lpss: Publish CSRT table for LPSS devices

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
soc/intel/baytrail/scc: Program PCI interrupts for SCC devices

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
src/mainboard/intel/minnowmax/irqroute.h: Add missing OTG IRQ

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
src/mainboard/intel/minnowmax/gpio.c: Match Intel reference FW settings

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
soc/intel/baytrail/southcluster.c: Use generic IOAPIC init API

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
soc/intel/baytrail: Postpone setting SSC ACPI mode to APM handler

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
soc/intel/baytrail: Allow disabling SERIRQ

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
mainboard/intel/minnowmax/devicetree.cb: Put SSC and LPSS into ACPI mode

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
device/pciexp_device.c: Use disable_pcie_aspm to disable ASPM too

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
soc/intel/baytrail/Makefile.mk: Check if MANIFESTS region start is correct

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
configs/config.intel_minnowmax_sb: Disable SMM BWP

SMM BWP does not work properly yet.

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
configs/config.intel_minnowmax: Bump to rc3

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 commented on src/soc/intel/baytrail/Kconfig:
```diff
@@ -255,16 +272,49 @@ config TXE_SB_SBM_MANIFEST_SVN
-# Put the microcode right after the manifest, so it will be verified by IBB
-# and accessible without walkcbfs ASM routine if CBFS_VERIFICATION is enabled.
-config CPU_MICROCODE_CBFS_LOC
+# This addresses will be used by BSP to locate microcode in bootblock before
```
```suggestion
# This address will be used by BSP to locate microcode in bootblock before
```

miczyg1 committed:
soc/intel/baytrail/Kconfig: Fix grammar

miczyg1 committed:
src/soc/intel/baytrail/Makefile.mk: Remove excessive comma

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
src/soc/intel/baytrail/Kconfig: Fix typo

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
src/soc/intel/baytrail/txei.c: Do not break printk string

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 commented on src/soc/intel/baytrail/txei.c:
```diff
@@ -0,0 +1,1039 @@
+
+	/* Write the body in whole slots */
+	for (i = 1; i < pend_slots; i++, *p++) {
+		write_slot(i, *p);
```
It may often be out of bound read, depending on the message length. If it is not divisible by 4, then some additional bytes may be sent at the end of message. However, ME/TXE should not care, because the real length of the message is in the message header. But we still must write 32bit slots... Of course we may always check for the message length, reallocate and copy it to a new buffer that is divisible by 4 and appended with zeros if necessary. However, the original source of CSE HECI `src/soc/intel/common/block/cse/cse.c` does not do it.

miczyg1 replied:
https://github.com/Dasharo/coreboot/commit/6a4524e4130d3cf460d9aa66859d1f9ff2af7aa2

This is not exactly what you wanted. Although like I mentioned above, it is based on the CSE HECI from common blocks, which does the same. And TXE/CSE should not care about bytes after the message size.

miczyg1 committed:
src/soc/intel/baytrail/txei.c: Refactor slot calculation

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 commented on src/soc/intel/baytrail/Kconfig:
```diff
@@ -255,16 +272,49 @@ config TXE_SB_SBM_MANIFEST_SVN
 
-endif
+# Ensure the bootblock does not overlap the microcode. We have 0x6400 bytes
+# of space after microcode till the ned of ROM. Make a small gap of 256 bytes
```
c23fdbee041001b49966caecd68e9cbe383895c1

miczyg1 commented on src/soc/intel/baytrail/txei.c:
```diff
@@ -0,0 +1,1039 @@
+/* get number of full 4-byte slots */
+static size_t bytes_to_slots(size_t bytes)
+{
+	return (bytes + (SLOT_SIZE - 1) + SLOT_SIZE) / SLOT_SIZE;
```
Reworked it to actually do what the comment does: 6a4524e4130d3cf460d9aa66859d1f9ff2af7aa2

Used ALIGN_UP like the original source of CSE HECI `src/soc/intel/common/block/cse/cse.c` does 

Also added comments near the loops where the slots are written to explain how it works.
